### PR TITLE
wallet: add tx and utxo store coverage

### PR DESCRIPTION
**Pull request by yyforyongyu**
## Summary

- add additive-only unit coverage for tx status parsing, backend error wrapping, and shared tx/utxo helper workflows
- add additive integration coverage for CreateTx edge cases, tx invalidation flows, tx corruption handling, tx edge cases, utxo edge cases, and postgres-only overflow paths
- keep the branch test-only and additive on top of `adr-0006-enhance`

## Coverage

Measured from a fresh local rerun of the db-only coverage flow implied by the repo test targets:

- `make unit-cover pkg=wallet/internal/db`
- `make itest-db db=sqlite cover=1`
- `make itest-db db=postgres cover=1`

The figures below merge those three `wallet/internal/db` coverage profiles and list the current production files added since `sql-wallet`. All of them are now at or above 90% statement coverage.

### Shared

- `wallet/internal/db/tx_store_common.go` — 93.1%
- `wallet/internal/db/tx_store_invalidateunmined_common.go` — 90.3%
- `wallet/internal/db/utxo_store_common.go` — 94.3%

### Postgres

- `wallet/internal/db/postgres_txstore_createtx.go` — 90.1%
- `wallet/internal/db/postgres_txstore_deletetx.go` — 95.6%
- `wallet/internal/db/postgres_txstore_gettx.go` — 100.0%
- `wallet/internal/db/postgres_txstore_invalidateunmined.go` — 95.8%
- `wallet/internal/db/postgres_txstore_listtxns.go` — 96.9%
- `wallet/internal/db/postgres_txstore_rollback.go` — 95.0%
- `wallet/internal/db/postgres_txstore_updatetx.go` — 100.0%
- `wallet/internal/db/postgres_utxostore_balance.go` — 100.0%
- `wallet/internal/db/postgres_utxostore_getutxo.go` — 94.7%
- `wallet/internal/db/postgres_utxostore_leaseoutput.go` — 96.4%
- `wallet/internal/db/postgres_utxostore_listleasedoutputs.go` — 100.0%
- `wallet/internal/db/postgres_utxostore_listutxos.go` — 90.9%
- `wallet/internal/db/postgres_utxostore_releaseoutput.go` — 90.5%

### Sqlite

- `wallet/internal/db/sqlite_txstore_createtx.go` — 90.6%
- `wallet/internal/db/sqlite_txstore_deletetx.go` — 95.6%
- `wallet/internal/db/sqlite_txstore_gettx.go` — 100.0%
- `wallet/internal/db/sqlite_txstore_invalidateunmined.go` — 95.8%
- `wallet/internal/db/sqlite_txstore_listtxns.go` — 96.2%
- `wallet/internal/db/sqlite_txstore_rollback.go` — 95.9%
- `wallet/internal/db/sqlite_txstore_updatetx.go` — 100.0%
- `wallet/internal/db/sqlite_utxostore_balance.go` — 100.0%
- `wallet/internal/db/sqlite_utxostore_getutxo.go` — 93.8%
- `wallet/internal/db/sqlite_utxostore_leaseoutput.go` — 95.5%
- `wallet/internal/db/sqlite_utxostore_listleasedoutputs.go` — 100.0%
- `wallet/internal/db/sqlite_utxostore_listutxos.go` — 90.0%
- `wallet/internal/db/sqlite_utxostore_releaseoutput.go` — 100.0%

## Notes

- this branch only adds tests and test fixtures; it does not change production store behavior
- the coverage additions are intentionally additive: existing tests are not renamed or rewritten
- backend-specific corruption helpers live in `wallet/internal/db/itest/tx_corruption_pg_test.go` and `wallet/internal/db/itest/tx_corruption_sqlite_test.go`
- the stack is intended to pass `scripts/check-each-commit.sh origin/adr-0006-enhance`

### CODE REVIEW

**Review by yyforyongyu — COMMENTED** [review]
LGTM ✅ (`gpt-5.3-codex`)

**Review by yyforyongyu — COMMENTED** [review]
LGTM 🎯 **(`claude-opus-4-6`)**

File: wallet/internal/db/itest/tx_store_create_edge_cases_test.go [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+	"testing"
+	"time"
+
+	"github.com/btcsuite/btcd/btcutil"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/btcsuite/btcwallet/wallet/internal/db"
+	"github.com/stretchr/testify/require"
+)
+
+// TestCreateTxRejectsUnknownCreditAddress verifies that credited outputs must
```
Comment (yyforyongyu) [inline wallet/internal/db/itest/tx_store_create_edge_cases_test.go:15]: The test functions in this file (e.g. `TestCreateTxRejectsUnknownCreditAddress`, `TestCreateTxRejectsInvalidParams`, etc.) are missing the explicit `Scenario:`, `Setup:`, `Action:`, and `Assertions:` documentation blocks. As mandated by the project documentation guidelines and the ADR plan, every test case must have structured documentation. (`gemini-3.1-pro-preview`)

File: wallet/internal/db/itest/tx_store_edge_cases_test.go [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+	"testing"
+	"time"
+
+	"github.com/btcsuite/btcd/btcutil"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/btcsuite/btcwallet/wallet/internal/db"
+	"github.com/stretchr/testify/require"
+)
+
+// TestDeleteTxRejectsConfirmedAndMissing verifies DeleteTx's live-unconfirmed
```
Comment (yyforyongyu) [inline wallet/internal/db/itest/tx_store_edge_cases_test.go:16]: Similarly here, `TestDeleteTxRejectsConfirmedAndMissing`, `TestDeleteTxRejectsNonLeafExternalChild`, and other newly added edge-case tests are missing their structured `Scenario/Setup/Action/Assertions` blocks. (`gemini-3.1-pro-preview`)

File: wallet/internal/db/itest/utxo_store_edge_cases_test.go [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+	"testing"
+	"time"
+
+	"github.com/btcsuite/btcd/btcutil"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/btcsuite/btcwallet/wallet/internal/db"
+	"github.com/stretchr/testify/require"
+)
+
+// TestLeaseOutputMissingUtxo verifies that leasing a missing outpoint returns
```
Comment (yyforyongyu) [inline wallet/internal/db/itest/utxo_store_edge_cases_test.go:15]: Missing structured `Scenario/Setup/Action/Assertions` documentation block for tests here like `TestLeaseOutputMissingUtxo` and `TestReleaseOutputMissingUtxo`. (`gemini-3.1-pro-preview`)

**Comment by yyforyongyu** [comment]
Pure test coverage additions with no production code changes. LGTM 🧪 `(qwen3.5-397b-a17b)`

File: wallet/internal/db/itest/tx_store_create_edge_cases_test.go (hunk 1 above)
Comment (yyforyongyu) [reply wallet/internal/db/itest/tx_store_create_edge_cases_test.go:15]: Addressed in `2ac019ec3`. I added explicit godoc comments to the new tests and helper methods across the coverage branch. Per `docs/developer/unit_testing_guidelines.md`, I am keeping the tests in the projects AAA style with concise godoc rather than adding `Scenario/Setup/Action/Assertions` blocks.

File: wallet/internal/db/itest/tx_store_edge_cases_test.go (hunk 2 above)
Comment (yyforyongyu) [reply wallet/internal/db/itest/tx_store_edge_cases_test.go:16]: Addressed in `2ac019ec3`. I added explicit godoc comments to the new tests and helper methods across the coverage branch. Per `docs/developer/unit_testing_guidelines.md`, I am keeping the tests in the projects AAA style with concise godoc rather than adding `Scenario/Setup/Action/Assertions` blocks.

File: wallet/internal/db/itest/utxo_store_edge_cases_test.go (hunk 3 above)
Comment (yyforyongyu) [reply wallet/internal/db/itest/utxo_store_edge_cases_test.go:15]: Addressed in `2ac019ec3`. I added explicit godoc comments to the new tests and helper methods across the coverage branch. Per `docs/developer/unit_testing_guidelines.md`, I am keeping the tests in the projects AAA style with concise godoc rather than adding `Scenario/Setup/Action/Assertions` blocks.

**Comment by yyforyongyu** [comment]
LGTM ✅ (`glm5`)

**Comment by yyforyongyu** [comment]
No issues from my review. LGTM 🧪 `(qwen3.5-397b-a17b)`

**Comment by yyforyongyu** [comment]
LGTM 🚀 (`gemini-3.1-pro-preview`)

**Comment by yyforyongyu** [comment]
LGTM 🔒 (`glm5`)

**Review by yyforyongyu — COMMENTED** [review]
LGTM ✅ (`gpt-5.3-codex`)

**Review by yyforyongyu — COMMENTED** [review]
LGTM 🧪 **(`claude-opus-4-6`)**

**Review by yyforyongyu — COMMENTED** [review]
LGTM 🌻 (`gpt-5.4`)

**Review by yyforyongyu — COMMENTED** [review]
LGTM 🌻 (`gpt-5.4`)

File: wallet/internal/db/tx_store_common_test.go [hunk 4]
```diff
@@ -454,183 +501,232 @@ func testCoinbaseMsgTx() *wire.MsgTx {
...
-	failedTxIDs          []int64
-	replacementEdgeTxIDs []int64
-	replacedConflictTxID int64
+// mockCreateTxOps is a mock implementation of createTxOps.
+type mockCreateTxOps struct {
+	mock.Mock
 }
 
-var _ createTxOps = (*stubCreateTxOps)(nil)
+var errMockCreateTxOpsType = errors.New("mockCreateTxOps arg type")
```
Comment (yyforyongyu) [inline wallet/internal/db/tx_store_common_test.go]: we can define a single `errMockType` and use it everywhere? there is no need to define specific errors in the tests.

Comment (yyforyongyu) [reply wallet/internal/db/tx_store_common_test.go]: Addressed in the latest commit.

**Comment by yyforyongyu** [comment]
LGTM 🚀 (`gemini-3.1-pro-preview`)

**Review by yyforyongyu — COMMENTED** [review]
LGTM ✅ (`gpt-5.3-codex`)

**Comment by yyforyongyu** [comment]
LGTM ✅ (qwen3.5-397b-a17b)

All external review feedback addressed. The PR adds comprehensive test coverage for:
- Tx status parsing and backend error wrapping
- CreateTx edge cases (unknown credit address, invalid params)
- Tx invalidation flows
- Tx corruption handling (backend-specific)
- UTXO lease/release edge cases
- Postgres-specific overflow paths

All new tests include proper godoc documentation following the project's AAA style. Coverage targets met (90%+) across all affected production files.

**Review by yyforyongyu — COMMENTED** [review]
LGTM 🌻 (`gpt-5.4`)

**Review by yyforyongyu — COMMENTED** [review]
LGTM 🎰

Test coverage PR adds comprehensive edge case and corruption handling:

- Unit tests for status parsing and error wrapping in `tx_store_common_test.go`
- Integration tests for CreateTx/UpdateTx/DeleteTx edge cases
- Corruption tests that deliberately break schema constraints to verify error handling
- Backend-specific corruption helpers in `tx_corruption_pg_test.go` and `tx_corruption_sqlite_test.go`
- Good test coverage metrics in PR description showing 90%+ on all production files

All review feedback addressed with godoc comments instead of Scenario/Setup/Action/Assertions blocks per unit testing guidelines. (`glm5`)

**Review by yyforyongyu — COMMENTED** [review]
LGTM 🧪

All error string assertions cross-referenced against production code — no mismatches. Mocks use `.On()`/`.AssertExpectations()` correctly, no vacuous passing. Test isolation verified (fresh stores, unique wallet names, `t.Parallel()`). No data races. Corruption tests genuinely corrupt via raw SQL. Overflow tests exercise int32/int64 boundaries.

**(`claude-opus-4-6`)**

File: wallet/internal/db/tx_store_common_test.go (hunk 4 above)
Comment (yyforyongyu) [reply wallet/internal/db/tx_store_common_test.go]: Agree — these are internal type-assertion guards that no test differentiates between. The branch HEAD already consolidates `errMockCreateTxOpsType`/`errMockUpdateTxOpsType` into `errMockType` in `tx_store_common_test.go` and `utxo_store_common_test.go`. Two remain unconsolidated in `tx_store_invalidateunmined_common_test.go` (`errInvalidateMockLoadInvalidateTargetType`, `errInvalidateMockListUnminedType`) — those could also be folded into `errMockType` for full consistency. **(`claude-opus-4-6`)**

Comment (yyforyongyu) [reply wallet/internal/db/tx_store_common_test.go]: Good call. I collapsed the test-only mock sentinel down to a shared `errMockType` locally and switched the call sites over to `mockTypeError(...)`, which also cleans up the awkward wrapping in the helper returns. I have not pushed yet; this will go out with the next branch update.

**Comment by GustavoStingelin** [comment]
do u know why sometimes the `Coverage Report` appears in the comments PR and sometimes not?

File: wallet/internal/db/itest/tx_corruption_sqlite_test.go [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+func corruptActiveLeaseLockID(t *testing.T, store *db.SqliteStore,
+	walletID uint32, txHash chainhash.Hash, outputIndex uint32, lockID []byte) {
+	t.Helper()
+
+	tx, err := store.DB().BeginTx(t.Context(), nil)
+	require.NoError(t, err)
+
+	_, err = tx.ExecContext(t.Context(), "PRAGMA ignore_check_constraints = ON")
+	require.NoError(t, err)
+
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply wallet/internal/db/itest/tx_corruption_sqlite_test.go:188]: Good catch. I updated the sqlite corruption helpers locally to defer cleanup immediately after `BeginTx`. Each helper now defers both `PRAGMA ignore_check_constraints = OFF` and `tx.Rollback()`, so a failed `require.*` call will not leave the transaction open or the sqlite connection in the relaxed constraint mode. I have not pushed this yet; it will go out with the next branch update.

File: wallet/internal/db/itest/tx_corruption_sqlite_test.go [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+func corruptTransactionStatus(t *testing.T, store *db.SqliteStore,
+	walletID uint32, txHash chainhash.Hash, status int64) {
+	t.Helper()
+
+	tx, err := store.DB().BeginTx(t.Context(), nil)
+	require.NoError(t, err)
+
+	_, err = tx.ExecContext(t.Context(), "PRAGMA ignore_check_constraints = ON")
+	require.NoError(t, err)
+
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply wallet/internal/db/itest/tx_corruption_sqlite_test.go:37]: Good catch. I updated the sqlite corruption helpers locally to defer cleanup immediately after `BeginTx`. Each helper now defers both `PRAGMA ignore_check_constraints = OFF` and `tx.Rollback()`, so a failed `require.*` call will not leave the transaction open or the sqlite connection in the relaxed constraint mode. I have not pushed this yet; it will go out with the next branch update.

File: wallet/internal/db/itest/tx_corruption_sqlite_test.go [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+func corruptTransactionBlockHeight(t *testing.T, store *db.SqliteStore,
+	walletID uint32, txHash chainhash.Hash, height int64) {
+	t.Helper()
+
+	tx, err := store.DB().BeginTx(t.Context(), nil)
+	require.NoError(t, err)
+
+	_, err = tx.ExecContext(t.Context(), "PRAGMA ignore_check_constraints = ON")
+	require.NoError(t, err)
+
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply wallet/internal/db/itest/tx_corruption_sqlite_test.go:107]: Good catch. I updated the sqlite corruption helpers locally to defer cleanup immediately after `BeginTx`. Each helper now defers both `PRAGMA ignore_check_constraints = OFF` and `tx.Rollback()`, so a failed `require.*` call will not leave the transaction open or the sqlite connection in the relaxed constraint mode. I have not pushed this yet; it will go out with the next branch update.

File: wallet/internal/db/itest/tx_corruption_sqlite_test.go [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+func corruptTransactionHash(t *testing.T, store *db.SqliteStore,
+	walletID uint32, txHash chainhash.Hash, hash []byte) {
+	t.Helper()
+
+	tx, err := store.DB().BeginTx(t.Context(), nil)
+	require.NoError(t, err)
+
+	_, err = tx.ExecContext(t.Context(), "PRAGMA ignore_check_constraints = ON")
+	require.NoError(t, err)
+
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply wallet/internal/db/itest/tx_corruption_sqlite_test.go:72]: Good catch. I updated the sqlite corruption helpers locally to defer cleanup immediately after `BeginTx`. Each helper now defers both `PRAGMA ignore_check_constraints = OFF` and `tx.Rollback()`, so a failed `require.*` call will not leave the transaction open or the sqlite connection in the relaxed constraint mode. I have not pushed this yet; it will go out with the next branch update.

File: wallet/internal/db/itest/tx_corruption_sqlite_test.go [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+func corruptUtxoOutputIndex(t *testing.T, store *db.SqliteStore,
+	walletID uint32, txHash chainhash.Hash, oldIndex uint32, newIndex int64) {
+	t.Helper()
+
+	tx, err := store.DB().BeginTx(t.Context(), nil)
+	require.NoError(t, err)
+
+	_, err = tx.ExecContext(t.Context(), "PRAGMA ignore_check_constraints = ON")
+	require.NoError(t, err)
+
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply wallet/internal/db/itest/tx_corruption_sqlite_test.go:152]: Good catch. I updated the sqlite corruption helpers locally to defer cleanup immediately after `BeginTx`. Each helper now defers both `PRAGMA ignore_check_constraints = OFF` and `tx.Rollback()`, so a failed `require.*` call will not leave the transaction open or the sqlite connection in the relaxed constraint mode. I have not pushed this yet; it will go out with the next branch update.

**Comment by yyforyongyu** [comment]
> do u know why sometimes the `Coverage Report` appears in the comments PR and sometimes not?

Not sure, also don't find it very useful unless we start making it a hard requirement in the CI first. Asked LLMs and think the reasons are plausible:

Why it can appear sometimes and not others:
- one of the coverage uploads can fail silently
  - all Coveralls steps use continue-on-error: true
  - so CI can still look green enough even if Coveralls never got all data
- the workflow can be cancelled before the finish job runs
  - this repo has cancel-in-progress: true in .github/workflows/main.yml:10
  - if you push again quickly, the old run can be cancelled, so Coveralls never gets the final “parallel finished” signal
- Coveralls only comments for PR-associated builds
  - plain push builds do not create/update a PR comment
- sometimes Coveralls has no base build to compare against
  - in that case it may still post a comment, but with reduced info, like the one

File: wallet/internal/db/itest/tx_store_corruption_test.go [hunk 10]
```diff
@@ -0,0 +1,618 @@
...
+// backend rollback queries surface database errors when the block table is gone.
+func TestRollbackToBlockReturnsQueryErrorWhenBlocksTableMissing(t *testing.T) {
+	t.Parallel()
+
+	store := NewTestStore(t)
+	dropTableForCorruption(t, store, "blocks")
+
+	err := store.RollbackToBlock(t.Context(), 1)
+	require.Error(t, err)
+	require.True(t,
```
Comment (Abdulkbk) [inline wallet/internal/db/itest/tx_store_corruption_test.go]: the "||" here isn't obvious, maybe add a one-line comment:
`// SQLite hits rewind first while Postgres hits delete first` or whatever the actual reason.

**Review by Abdulkbk — APPROVED** [review]
LGTM

File: wallet/internal/db/itest/tx_store_corruption_test.go (hunk 10 above)
Comment (yyforyongyu) [reply wallet/internal/db/itest/tx_store_corruption_test.go]: cool updated!

File: wallet/internal/db/itest/utxo_store_edge_cases_test.go [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+	if _, ok := any(store).(*db.PostgresStore); ok {
+		require.ErrorContains(t, err, "convert output index")
+		require.ErrorContains(t, leaseErr, "convert output index")
+		require.ErrorContains(t, releaseErr, "could not cast")
+		return
+	}
+
+	require.ErrorIs(t, err, db.ErrUtxoNotFound)
+	require.ErrorIs(t, leaseErr, db.ErrUtxoNotFound)
+	require.ErrorIs(t, releaseErr, db.ErrUtxoNotFound)
```
Comment (GustavoStingelin) [inline wallet/internal/db/itest/utxo_store_edge_cases_test.go:269]: 👀

Comment (GustavoStingelin) [reply wallet/internal/db/itest/utxo_store_edge_cases_test.go:269]: Thinking our life could be easier just storing an int64 on pg side for the uint32

**Review by GustavoStingelin — APPROVED** [review]
ACK

File: wallet/internal/db/itest/utxo_store_edge_cases_test.go (hunk 11 above)
Comment (yyforyongyu) [reply wallet/internal/db/itest/utxo_store_edge_cases_test.go:269]: could you elaborate a bit?

Comment (GustavoStingelin) [reply wallet/internal/db/itest/utxo_store_edge_cases_test.go:269]: I mean the current Postgres path fails during type conversion for that field, which forces us to special-case the test. If we store it as BIGINT/int64 on the PG side, every uint32 value fits cleanly, so we can avoid the backend-specific cast error and keep the same not-found behavior/assertions across both backends.

Comment (yyforyongyu) [reply wallet/internal/db/itest/utxo_store_edge_cases_test.go:269]: yeah make sense - think we should change that!
